### PR TITLE
feat(manager/git-submodule): add authentification support

### LINKED ISSUE #10149: Issues cloning private git-submodules
**How are you running Renovate?**

- [x] WhiteSource Renovate hosted app on github.com
- [ ] Self hosted

If using the hosted app, please skip to the next section.
Otherwise, if self-hosted, please complete the following:

Please select which platform you are using:

- [x] Azure DevOps (dev.azure.com)
- [ ] Azure DevOps Server
- [ ] Bitbucket Cloud (bitbucket.org)
- [ ] Bitbucket Server
- [ ] Gitea
- [ ] github.com
- [ ] GitHub Enterprise Server
- [ ] gitlab.com
- [ ] GitLab self-hosted

Renovate version: Unknown, using GitHub Applications

**Describe the bug**

The beta git-submodules seems to have an issue cloning repos in the same org but private, hopefully I'm not doing something stupid! I've included my config to hopefully help as well.

**Relevant debug logs**

<details><summary>Click me to see logs</summary>

```
DEBUG: Cloning git submodule at monitoring
WARN: Unable to initialise git submodule at monitoring
{
  "err": {
    "task": {
      "commands": [
        "submodule",
        "update",
        "--init",
        "monitoring"
      ],
      "format": "utf-8"
    },
    "message": "Submodule 'kubernetes-monitoring' (https://github.com/polr-software/kubernetes-monitoring) registered for path 'monitoring'\nCloning into '/mnt/renovate/gh/polr-software/kubernetes/monitoring'...\nfatal: could not read Username for 'https://github.com': No such device or address\nfatal: clone of 'https://github.com/polr-software/kubernetes-monitoring' into submodule path '/mnt/renovate/gh/polr-software/kubernetes/monitoring' failed\nFailed to clone 'monitoring'. Retry scheduled\nCloning into '/mnt/renovate/gh/polr-software/kubernetes/monitoring'...\nfatal: could not read Username for 'https://github.com': No such device or address\nfatal: clone of 'https://github.com/polr-software/kubernetes-monitoring' into submodule path '/mnt/renovate/gh/polr-software/kubernetes/monitoring' failed\nFailed to clone 'monitoring' a second time, aborting\n",
    "stack": "Error: Submodule 'kubernetes-monitoring' (https://github.com/polr-software/kubernetes-monitoring) registered for path 'monitoring'\nCloning into '/mnt/renovate/gh/polr-software/kubernetes/monitoring'...\nfatal: could not read Username for 'https://github.com': No such device or address\nfatal: clone of 'https://github.com/polr-software/kubernetes-monitoring' into submodule path '/mnt/renovate/gh/polr-software/kubernetes/monitoring' failed\nFailed to clone 'monitoring'. Retry scheduled\nCloning into '/mnt/renovate/gh/polr-software/kubernetes/monitoring'...\nfatal: could not read Username for 'https://github.com': No such device or address\nfatal: clone of 'https://github.com/polr-software/kubernetes-monitoring' into submodule path '/mnt/renovate/gh/polr-software/kubernetes/monitoring' failed\nFailed to clone 'monitoring' a second time, aborting\n\n    at Object.action (/home/ubuntu/renovateapp/node_modules/simple-git/src/lib/plugins/error-detection.plugin.js:30:33)\n    at PluginStore.exec (/home/ubuntu/renovateapp/node_modules/simple-git/src/lib/plugins/plugin-store.js:21:33)\n    at /home/ubuntu/renovateapp/node_modules/simple-git/src/lib/runners/git-executor-chain.js:95:45\n    at new Promise (<anonymous>)\n    at GitExecutorChain.handleTaskData (/home/ubuntu/renovateapp/node_modules/simple-git/src/lib/runners/git-executor-chain.js:93:16)\n    at GitExecutorChain.<anonymous> (/home/ubuntu/renovateapp/node_modules/simple-git/src/lib/runners/git-executor-chain.js:77:46)\n    at Generator.next (<anonymous>)\n    at fulfilled (/home/ubuntu/renovateapp/node_modules/simple-git/src/lib/runners/git-executor-chain.js:5:58)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:93:5)"
  }
}
DEBUG: Cloning git submodule at mqtt
```

</details>

**Have you created a minimal reproduction repository?**

Please read the [minimal reproductions documentation](https://github.com/renovatebot/renovate/blob/main/docs/development/minimal-reproductions.md) to learn how to make a good minimal reproduction repository.

- [ ] I have provided a minimal reproduction repository
- [ ] I don't have time for that, but it happens in a public repository I have linked to
- [ ] I don't have time for that, and cannot share my private repository
- [x] The nature of this bug means it's impossible to reproduce publicly

**Additional context**

I looked to create a minimal reproduction repository, unfortunately I think this is an issue directly related to private repositories and auth

Below is the config I'm using:
```
{
  "extends": [
    "config:base"
  ],
  "kubernetes": {
    "fileMatch": ["\\.yaml$"]
  },
  "git-submodules": {
    "enabled": true
  },
  "reviewers": ["team:smart-badge-platform"]
}
```

### PR DESCRIPTION
**Pull request by Shegox**
## Changes:

Add authentication to `simple-git` using the new GIT Environment variables option.

## Context:

This is only a PoC to share a potential solution; don't merge/review yet.

Fixes #10149

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/manager/git-submodules/update.ts [hunk 1]
```diff
@@ -1,17 +1,31 @@
-import Git from 'simple-git';
+import simpleGit from 'simple-git';
```
Comment (viceice) [inline lib/manager/git-submodules/update.ts:1]: Why rename?

File: lib/manager/git-submodules/update.ts [hunk 2]
```diff
@@ -21,3 +35,61 @@ export default async function updateDependency({
...
+
+  if (githubToken?.token) {
+    environmentVariables = getGitAuthenticatedEnvironmentVariables(
+      'https://github.com/',
+      githubToken
+    );
+  }
+
+  // get extra host rules for other git-based Go Module hosts
+  const hostRules = getAll() || [];
```
Comment (viceice) [inline lib/manager/git-submodules/update.ts:56]: We shouldn't use all hostrules. Instead we should read all submodule urls and check for matching hostrules only.

**Review by viceice — CHANGES_REQUESTED** [review]

**Comment by PhilipAbed** [comment]
Hi @Shegox this is very old, is this still relevant?

**Comment by rarkins** [comment]
Closing due to staleness

**Pull request closed without merging**